src/lib.rs

```diff
@@ -193,9 +185,26 @@ mod windows {
 }
 
 #[cfg(unix)]
-pub use unix::read_password_with_reader;
+pub use unix::read_password_from_stdin;
 #[cfg(windows)]
-pub use windows::read_password_with_reader;
+pub use windows::read_password_from_stdin;
+
+/// Reads a password from anything that implements BufRead
+pub fn read_password_with_reader<T>(source: Option<T>) -> ::std::io::Result<String>
+    where T: ::std::io::BufRead {
+    match source {
+        Some(mut reader) => {
+            let mut password = String::new();
+            if let Err(err) = reader.read_line(&mut password) {
```

Comment:
This seems like the library would not hide the password if you pass `stdin()` to `read_password_with_reader`, with something like:

```rust
let stdin = std::io::stdin();
read_password_with_reader(Some(stdin));
```
(sorry if the syntax is a bit off, I haven't been coding much Rust lately)

We introduced `read_password_with_reader` specifically for people who dynamically inject a reader into a program, which allows unit testing the program in pure Rust, without calling commands from a shell. So they might do this:
```rust
fn main() {
    let reader = std::io::stdin();

    run_program(reader);
}

#[test]
fn test_program() {
    let test_reader = get_test_reader();
    run_program(test_reader);

    // some assertions here
    // ...
}
```

This means that `read_password_from_with_reader()`, should, I think, keep hiding the password.

What do you think ?

Reply:
Also, given that, as I understand it, `rpassword` currently is completely unable to read from a tty, maybe just adding the new function from #26 without changing existing functions would allow to work for your need without breaking existing code, right ? Given the small scope of the library, it's fine is some code is duplicated IMO.

Reply:
> This seems like the library would not hide the password if you pass `stdin()` to `read_password_with_reader`, with something like:
> 
> ```rust
> let stdin = std::io::stdin();
> read_password_with_reader(Some(stdin));
> ```
> 
> (sorry if the syntax is a bit off, I haven't been coding much Rust lately)
> 
> We introduced `read_password_with_reader` specifically for people who dynamically inject a reader into a program, which allows unit testing the program in pure Rust, without calling commands from a shell. So they might do this:
> 
> ```rust
> fn main() {
>     let reader = std::io::stdin();
> 
>     run_program(reader);
> }
> 
> #[test]
> fn test_program() {
>     let test_reader = get_test_reader();
>     run_program(test_reader);
> 
>     // some assertions here
>     // ...
> }
> ```

Shouldn't they do


> 
> This means that `read_password_from_with_reader()`, should, I think, keep hiding the password.
> 
> What do you think ?

Reply:
> Also, given that, as I understand it, `rpassword` currently is completely unable to read from a tty, maybe just adding the new function from #26 without changing existing functions would allow to work for your need without breaking existing code, right ? Given the small scope of the library, it's fine is some code is duplicated IMO.

So you wanting to be backwards compatible is not helpful with regard to creating a good API for the crate.  IMHO a crate should do one thing, and do it well.  As of now, your crate has four functions, all of which are variations on reading a password, however, neither of which does it securely, i.e. without accidentally reading from a redirected stdin, or without clobbering either stdout or stderr, either of which may be redirected, and hence may either not reach the user, interfere with a data stream, or both.

I'd very much prefer the crate to have one function, `get`, that securely reads a password from the console.

As for testing, I believe that handing in an io::Read is the wrong approach.  Instead, the program to be tested should be handed a function, either `rpassword::get`, or `my_mock_password_function`, both returning an `io::Result<String>`.